cmd/juju/user/login.go

```diff
@@ -132,142 +134,86 @@ func (c *loginCommand) Run(ctx *cmd.Context) error {
-	if err != nil {
+	// Find out details on the specified controller.
+	controllerDetails, err := store.ControllerByName(c.controllerName)
+	if err != nil && !errors.IsNotFound(err) {
 		return errors.Trace(err)
 	}
-	conn, err := newAPIConnection(params)
-	if err != nil {
-		return errors.Annotate(err, "creating API connection")
-	}
-	defer conn.Close()
 
-	accountDetails.LastKnownAccess = conn.ControllerAccess()
-	if err := store.UpdateAccount(controllerName, *accountDetails); err != nil {
-		return errors.Annotate(err, "failed to record temporary credential")
+	// Find out details of the controller domain if it's specified.
+	var conn api.Connection
+	var publicControllerDetails *jujuclient.ControllerDetails
+	var accountDetails *jujuclient.AccountDetails
+	// Fetch current details for the specified controller name so we
+	// can tell if the logged in user has changed.
+	oldAccountDetails, err := store.AccountDetails(c.controllerName)
+	if err != nil && !errors.IsNotFound(err) {
+		return errors.Trace(err)
 	}
-	ctx.Infof("You are now logged in to %q as %q.", controllerName, userTag.Id())
-	return nil
-}
-
-func (c *loginCommand) controllerLogin(ctx *cmd.Context, store jujuclient.ClientStore) error {
-	controllerHost := c.userOrDomain
-	if !strings.Contains(controllerHost, ".") {
-		// TODO(rog) provide a way of avoiding this external network access.
-		controllerHost1, err := c.getKnownControllerDomain(controllerHost)
-		if errors.IsNotFound(err) {
-			return errors.Errorf("%q is not a known public controller", controllerHost)
+	if c.domain != "" {
+		conn, publicControllerDetails, accountDetails, err = c.publicControllerLogin(ctx, c.domain, oldAccountDetails)
+		if err != nil {
+			return errors.Annotatef(err, "cannot log into %q", c.domain)
 		}
+	} else if controllerDetails == nil {
+		// No controller found and no domain specified - we
+		// have no idea where we should be logging in.
+		return errors.Errorf("domain not specified and controller does not exist")
+	} else {
+		conn, accountDetails, err = c.existingControllerLogin(ctx, store, c.controllerName, oldAccountDetails)
 		if err != nil {
-			return errors.Annotatef(err, "could not determine controller domain")
+			return errors.Annotatef(err, "cannot log into controller %q", c.controllerName)
 		}
-		controllerHost = controllerHost1
 	}
-	if c.controllerName == "" {
-		// No explicitly specified controller name, so
-		// derive it from the domain.
-		if strings.Contains(c.userOrDomain, ":") {
-			return errors.Errorf("cannot use %q as controller name - use -c flag to choose a different one", c.userOrDomain)
+	defer conn.Close()
+	if controllerDetails != nil && publicControllerDetails != nil && controllerDetails.ControllerUUID != publicControllerDetails.ControllerUUID {
+		// The domain we're trying to log into doesn't match the
+		// existing controller.
+		return errors.Errorf(`
+controller at %q does not match existing controller.
```

Comment:
An end user would not really grok this message I don't think.
How about something like 

Connection details for controller %q have changed.
You can remove the outdated controller details using 
"juju unregister %s".
To login to a different controller, use the -c flag to specify another name"